### PR TITLE
Sfi 76 Retrieve and update an incomplete application via the organisation page

### PR DESCRIPTION
**Pull request by simonnineteen90**
https://eaflood.atlassian.net/browse/SFI-76

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ffc-sfi-apply-web",
-  "version": "1.39.1",
```
Comment (marctemp) [inline package.json:3]: Maybe bump the version to `1.40.0` as its a functional change

File: app/agreement/create.js [hunk 2]
```diff
@@ -6,6 +6,7 @@ const AGREEMENT = {
   agreementNumber: undefined,
   confirmed: false,
   submitted: false,
+  statusId: 1,
```
Comment (marctemp) [inline app/agreement/create.js:9]: Could we create a data model map for statusIds to eliminate magic numbers?

I have no idea what a statusId or 1 means for e.g.

File: app/routes/schemas/agreementNumber.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+const Joi = require('joi')
```
Comment (marctemp) [inline app/routes/schemas/agreementNumber.js:1]: Files are named in kebab-case rather than camelCase

File: test/integration/narrow/routes/start-application.test.js [hunk 4]
```diff
@@ -2,7 +2,9 @@ describe('start-application route', () => {
   jest.mock('ffc-messaging')
   jest.mock('../../../../app/plugins/crumb')
   jest.mock('../../../../app/cache')
+  jest.mock('../../../../app/agreement/get')
   const mockCache = require('../../../../app/cache')
```
Comment (marctemp) [inline test/integration/narrow/routes/start-application.test.js:6]: Consider adding whitespace

File: app/routes/start-application.js [hunk 5]
```diff
@@ -23,22 +29,38 @@ module.exports = [{
   path: '/start-application',
   options: {
     validate: {
-      payload: schema
+      payload: Joi.object()
+        .concat(sbiSchema)
+        .concat(agreementNumberSchema)
```
Comment (marctemp) [inline app/routes/start-application.js:34]: Oh nice, did not know you could do this 👍

File: app/routes/start-application.js [hunk 6]
```diff
@@ -23,22 +29,38 @@ module.exports = [{
...
+      if (!request.payload.agreementNumber) {
         agreement = create()
-        // Need to include caller Id as part of agreement to support downstream services
-        // Expect to remove once Defra Identity integrated
-        agreement.callerId = callerId
-        agreement.crn = crn
-      }
+      } else {
+        const incompleteApplications = await getBySbi(sbi)
+        const agreements = incompleteApplications.agreements
```
Comment (marctemp) [inline app/routes/start-application.js:46]: Consider `const { agreements } = incompleteApplications` for consistency

(I also thought doing this eliminated silent issues as `const { notInTheObject } = anObject` would error saying does not exist but does not , returns undefined which then when you try to expand errors, sad times 😢 )

File: app/routes/start-application.js (hunk 5 above)
Comment (marctemp) [inline app/routes/start-application.js:34]: Should there not be a failAction if the schema fails?

File: app/views/start-application.njk [hunk 7]
```diff
@@ -44,4 +45,62 @@
...
+              {% else %}
+                 <td class="govuk-table__cell">
+                  <strong class="govuk-tag govuk-tag--grey"> {{ agreement.status.status }} </strong>
+                </td>
+              {% endif %}
+
+              <td class="govuk-table__cell">
+                {% if agreement.status.status === 'In Progress' %}
+                  <form method="POST" action="/start-application" novalidate>
+                    <input type="hidden" name="crumb" value="{{crumb}}"/>
```
Comment (marctemp) [inline app/views/start-application.njk:86]: consider `value="{{ crumb }}"` for consistency

File: test/integration/narrow/routes/start-application.test.js [hunk 8]
```diff
@@ -22,6 +24,31 @@ describe('start-application route', () => {
     address: 'address1, address2, address3, postalCode'
   }]
 
+  const applications = {
+    agreements: [
+      {
+        agreementNumber: 'SFIA123456',
+        agreementData: {
+          land: 'land',
+          funding: 'funding',
```
Comment (marctemp) [inline test/integration/narrow/routes/start-application.test.js:33]: Interesting approach to mocking values that you do not directly care about in the tests -- be interesting to hear what other make of this and whether a mock_cache or mock_models for some entities should exist and tests read it in?

if you remove these unrequired keys, I assume the programme breaks as somewhere a piece of code is accessing them but you are not querying their output, hence this functions correctly?

File: app/routes/start-application.js [hunk 9]
```diff
@@ -13,7 +17,9 @@ module.exports = [{
       const sbi = request.query?.sbi ?? agreement?.organisation?.sbi
       if (sbi) {
```
Comment (marctemp) [inline app/routes/start-application.js:18]: I think there's edge cases not handled here:
- what if SBI is invalid (too short, not numbers, too long, etc is given, we have the SBI schema, just needs added)
- what if SBI is valid but does not correspond to a organisation record?

Just an observation because this is outside the scope of this ticket and change.

**Comment by marctemp** [comment]
Great code guys @pxp420 and @simonnineteen90, looks great!

Only have some minor suggestions left; need to run the repo assa whole but I'm heading off now so will pick this back up on Monday -- enjoy your weekends!

File: app/routes/start-application.js [hunk 10]
```diff
@@ -23,22 +29,38 @@ module.exports = [{
...
+      await cache.reset(request)
+      if (!request.payload.agreementNumber) {
         agreement = create()
-        // Need to include caller Id as part of agreement to support downstream services
-        // Expect to remove once Defra Identity integrated
-        agreement.callerId = callerId
-        agreement.crn = crn
-      }
+      } else {
+        const incompleteApplications = await getBySbi(sbi)
```
Comment (steve-dickinson) [inline app/routes/start-application.js:45]: Instead of using getBySbi and having to check the array you can use getByAgreementNumber. Also you should be able to cut down on building up the agreement object:

```
      const sbi = parseInt(request.payload.sbi)
      const agreementNumber = request.payload.agreementNumber

      await cache.reset(request)
      if (!agreementNumber) {
        agreement = create()
        // Need to include caller Id as part of agreement to support downstream services
        // Expect to remove once Defra Identity integrated
        agreement.callerId = callerId
        agreement.crn = crn
        const selectedOrganisation = data.eligibleOrganisations.find(x => x.sbi === sbi)

        if (selectedOrganisation) {
          agreement.organisation = selectedOrganisation
        }
      } else {
        const incompleteApplications = await getByAgreementNumber(agreementNumber, sbi)

        if (!incompleteApplications) {
          return h.view('404')
        }

        agreement = incompleteApplications.agreementData
      }

      if (agreement.organisation) {
        await cache.update(request, { agreement })
        return h.redirect('/task-list')
      }

      return h.redirect('/eligible-organisation')
    }
```

File: app/views/start-application.njk [hunk 11]
```diff
@@ -44,4 +45,62 @@
...
+              <th scope="col" class="govuk-table__header"></th>
+            </tr>
+          </thead>
+          <tbody class="govuk-table__body">
+            {% for agreement in agreements|sort(true, false, 'updatedAtFormatted') %}
+            <tr class="govuk-table__row">
+              <td class="govuk-table__cell">{{ agreement.agreementNumber }}</td>
+              <td class="govuk-table__cell">{{ agreement.updatedAtFormatted }}</td>
+
+              {% if agreement.status.status === 'In Progress' %}
```
Comment (steve-dickinson) [inline app/views/start-application.njk:69]: Could build a macro to handle the status:

```
{% from "govuk/components/tag/macro.njk" import govukTag %}

{% macro taskLink(status) %}
  <strong class="govuk-tag app-task-list__tag {{ 'govuk-tag--yellow' if status !== 'In Progress'}} {{ 'govuk-tag--blue' if status === 'Submitted'}} {{ 'govuk-tag--grey' if status === ''}}">{{status}}</strong>
{% endmacro %}
```

**Review by steve-dickinson — CHANGES_REQUESTED** [review]

**Review by steve-dickinson — COMMENTED** [review]
My 2 cents, but good work :)

File: test/integration/narrow/routes/start-application.test.js (hunk 8 above)
Comment (pxp420) [reply test/integration/narrow/routes/start-application.test.js:34]: Yeah exactly, the values of those keys are not used during the test so they are just added as mocks to ensure the format of the object is correct. Can change it to a real value if needed.

File: app/routes/start-application.js (hunk 9 above)
Comment (pxp420) [reply app/routes/start-application.js]: I think both this scenario, and the one where there is nothing displayed as an error when POSTing are not covered in this ticket as it doesn't state what the behaviour should be, what error message to be displayed, etc.

File: app/routes/start-application.js (hunk 5 above)
Comment (pxp420) [reply app/routes/start-application.js]: Similar to the comment below. Not mentioned in the ticket and before working on this, there wasn't a failAction even though the schema was already checking for the SBI. It redirects the user to a 400 page if any validation fails which I think is good enough for now.

File: app/views/start-application.njk (hunk 7 above)
Comment (pxp420) [reply app/views/start-application.njk]: changed

File: app/routes/start-application.js (hunk 6 above)
Comment (pxp420) [reply app/routes/start-application.js]: changed

File: test/integration/narrow/routes/start-application.test.js (hunk 4 above)
Comment (pxp420) [reply test/integration/narrow/routes/start-application.test.js:7]: added

File: app/routes/schemas/agreementNumber.js (hunk 3 above)
Comment (pxp420) [reply app/routes/schemas/agreementNumber.js]: changed

File: app/agreement/create.js (hunk 2 above)
Comment (pxp420) [reply app/agreement/create.js]: added a map for consistency

File: package.json (hunk 1 above)
Comment (pxp420) [reply package.json:3]: changed

File: app/routes/start-application.js (hunk 10 above)
Comment (pxp420) [reply app/routes/start-application.js]: I think the reason why we used getBySbi is because we need the list of agreementNumbers so that we can display the table. If we used getByAgreementNumber we would lose that information.

Also thanks for this:
`agreement = incompleteApplications.agreementData`

we didn't think was an easier way!

File: app/views/start-application.njk (hunk 11 above)
Comment (pxp420) [reply app/views/start-application.njk:69]: added the macro

File: app/agreement/submit.js [hunk 12]
```diff
@@ -2,15 +2,18 @@ const cache = require('../cache')
...
   await save(request)
   const { agreement } = await cache.get(request)
   const { agreementNumber, organisation } = agreement
 
+  const submitted = status.find(x => x.name === 'Submitted')
+
   const url = '/agreement/submit'
   await post(url, { agreementNumber, sbi: organisation.sbi })
-  await cache.update(request, { agreement: { submitted: true } })
+  await cache.update(request, { agreement: { submitted: true, statusId: submitted } })
```
Comment (marctemp) [inline app/agreement/submit.js:16]: Think this needs to be `submitted.id` as done in `agreement/create.js`

File: app/agreement/status.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
+// module export
+// import into create.js
```
Comment (marctemp) [inline app/agreement/status.js:2]: Unneeded comments

**Review by marctemp — CHANGES_REQUESTED** [review]
Pipeline fails on Snyk stage, I had this on my last PR:

- remove package-lock.json and re-run npm update to recreate it

File: app/agreement/status.js (hunk 13 above)
Comment (pxp420) [reply app/agreement/status.js]: removed

File: app/agreement/submit.js (hunk 12 above)
Comment (pxp420) [reply app/agreement/submit.js]: great spot, thanks!

**Review by marctemp — APPROVED** [review]

File: app/routes/start-application.js (hunk 10 above)
Comment (pxp420) [reply app/routes/start-application.js]: Yeah, that is very true, sorry. I thought you were also suggesting it in the GET route. We used the `getByAgreementNumber` for POST in the end so I think it aligns with what you were referring to.

**Review by steve-dickinson — APPROVED** [review]
LGTM
